### DISCUSSION TITLE
config help: renovate.json with terraform and terragrut monorepo

### DISCUSSION THREAD

**Discussion by sajidk1** [renovatebot/renovate#11118] [Request Help] [answered]
Hi there 👋 

## Group PRs by path

I'm trying to configure renovate to group PRs by path.

You can see the repo, and my `renovate.json` here: https://github.com/sajid-khan-js/renovate-test. I'm using the GItHub app. I've left my config on `main` vanilla for now, but my last attempt is [here](https://github.com/sajid-khan-js/renovate-test/blob/2022dfcb97d03f831d86b2dd9acc1834eabbffde/renovate.json)

Essentially, I want a PR per path (e.g. `infra/dev/` or `modules/dummy_stack/`), the PRs should include all dependency updates for their respective paths.

With my last config attempt (linked above) renovate seems to be ignoring everything (resulting in no PRs)? What's weird is I did manage to get it to work for one path when my config looked like this:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "packageRules": [
    {
      "groupName": "Dev updates",
      "matchPaths": ["dev/"],
      "versioning": "regex:^((?<compatibility>.*)-v|v*)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$"
    }
  ]
}
```

You can see the resulting PR here: https://github.com/sajid-khan-js/renovate-test/pull/6

## Why isn't renovate bumping...

Even with a vanilla config, renovate seems to be ignoring certain dependencies

- Why isn't renovate bumping providers? See [here](https://github.com/sajid-khan-js/renovate-test/blob/2df70e927aef32f5edf45e5e4aae6957e81fb1f5/modules/dummy_stack/versions.tf#L7)
- Why isn't renovate bumping terraform versions where a pin like so is used `>= 0.14.4`, see file [here](https://github.com/sajid-khan-js/renovate-test/blob/2df70e927aef32f5edf45e5e4aae6957e81fb1f5/modules/dummy_module/versions.tf#L2)
- Why isn't renvoate bumping upstream modules in terragrunt. See [here](https://github.com/sajid-khan-js/renovate-test/blob/2df70e927aef32f5edf45e5e4aae6957e81fb1f5/infra/dev/deployment/terragrunt.hcl#L6)

You can see the resulting PRs that were open from the vanilla config here: https://github.com/sajid-khan-js/renovate-test/pulls

## Dynamic grouping

Hopefully my config just needs work and my use case is currently supported by renovate!

However, am I right in saying that I need to statically define each path? As in, there is no current renovate feature that automatically start grouping PRs if I added `modules/dummy_stack_2/` to my repo, i.e. nothing would happen until a `packageRule` was added to my `renovate.json`? #11092 seems to suggest this is the case.

**Comment by rarkins** [comment]
I'm seeing at least 4 separate questions here, each with its own config attempt. It's best if you can start with the problem you most want solved, with an example/config which is stripped down to reflect it.

**Reply by sajidk1** [reply]
Fair point @rarkins. I've updated the comment.

**Comment by rarkins** [comment, marked as answer]
What does the "packageRules with updates" log message say about your dependencies? It should list every one it detected and if there is any skip reason. I'd also recommend dropping the regex manager for now in case it's causing the dependency to be rejected.

**Reply by sajidk1** [reply]
Turns out this was me not understanding how renovate manages state (branches and PRs), I also ended up running renovate locally, so I could control when it runs.

I've updated the readme in my repo with my findings: https://github.com/sajid-khan-js/renovate-test

I've restored my original comment (with multiple queries), since my readme goes over them.

@rarkins Just one outstanding query for me then, please:

> However, am I right in saying that I need to statically define each path? As in, there is no current renovate feature that automatically starts grouping PRs if I added `modules/dummy_stack_2/` to my repo, i.e. nothing would happen until a `packageRule` was added to my `renovate.json`? #11092 seems to suggest this is the case.

**Reply by rarkins** [reply]
You can perhaps use templating with fields `parentDir` or `baseDir`. Please search this repo for previously issues or discussions for examples - I'm on mobile now so cannot easily

**Reply by sajidk1** [reply]
Thanks for the tip!

```json
    {
      "groupName": "{{parentDir}}",
      "additionalBranchPrefix": "{{parentDir}}-",
      "matchPaths": ["**/modules/**"]
    }
```

Did the trick for grouping by parent folder., but for grouping by common ancestor (e.g. `dev` or `prd` in the tree output below), I don't think there's anyway unless you statically define something.

```
├── infra
│   ├── dev
│   │   └── deployment
│   │       └── terragrunt.hcl
│   └── prd
│       └── deployment
│           └── terragrunt.hcl
├── modules
│   ├── dummy_module
│   │   ├── README.md
│   │   ├── main.tf
│   │   ├── outputs.tf
│   │   ├── variables.tf
│   │   └── versions.tf
│   └── dummy_stack
│       ├── README.md
│       ├── main.tf
│       ├── outputs.tf
│       ├── terraform.tfstate
│       ├── variables.tf
│       └── versions.tf
```

But for my use case that static definition is minimal and should not change (i.e they are env names). 

It was the modules dynamic grouping that was more important, since new modules will be added frequently.

Thanks for all your help, and thanks for making the awesome tool that is renovate!
